src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java

```diff
@@ -93,10 +105,46 @@ private JsonArray formatBatch(String topic) {
 
             JsonElement recordJSON = recordToJSON(record);
             JsonObject message = populateMetadata(topic, recordJSON);
-            batchRecords.add(message);
+            JsonArray currentBatchRecords = batchRecordsList.get(currentBatchListIndex);
+
+            try {
+                if (hasBatchOverflowed(message, currentBatchRecords)) {
+                    JsonArray emptyBatch = new JsonArray();
+
+                    emptyBatch.add(message);
+                    batchRecordsList.add(currentBatchListIndex++, emptyBatch);
+                } else {
+                    currentBatchRecords.add(message);
+                }
+
+            } catch (SizeLimitExceededException ex) {
+                log.error("Single message exceeds size limit.", ex);
+            }
         }
+        return batchRecordsList;
+    }
 
-        return batchRecords;
+    private boolean hasBatchOverflowed(JsonObject jsonRecord, JsonArray
+            batchRecords) throws SizeLimitExceededException {
+        String jsonRecordString = jsonRecord.toString();
+        UUID id = UUID.randomUUID();
+
+        batchRecords.add(jsonRecord);
+        String batchRecordsString = batchRecords.toString();
+        final byte[] batchRecordsStringBytes = batchRecordsString.getBytes(StandardCharsets.UTF_8);
+        batchRecords.remove(jsonRecord);
+
+        if (batchRecordsStringBytes.length / ONE_MEGABYTE >= MAXIMUM_BATCH_MEGABYTES) {
+            log.warn("Splitting batch because of size limits. Bytes of batch after new message was added: " + batchRecordsStringBytes.length + " id: " + id);
+            final byte[] jsonRecordStringBytes = jsonRecordString.getBytes(StandardCharsets.UTF_8);
+
+            if (batchRecords.size() == MINIMUM_LIST_LENGTH || jsonRecordStringBytes.length / ONE_MEGABYTE >= MAXIMUM_BATCH_MEGABYTES) {
+                throw new SizeLimitExceededException(String.format("Single message exceeds JSON size limit. " +
+                        "Truncated message: %s, id: %s", jsonRecordString.substring(MINIMUM_STRING_LENGTH, jsonRecordString.length() / TRUNCATE_DIVIDER), id));
+            }
+            return true;
+        }
```

Comment:
I think this could be slightly more efficient by storing a constant of the max batch size in bytes rather than needing to do division.